execution/stream_join.go

```diff
@@ -0,0 +1,448 @@
+		leftSource:       leftSource,
+		rightSource:      rightSource,
+		storage:          storage,
+		eventTimeField:   eventTimeField,
+		joinType:         joinType,
+		triggerPrototype: triggerPrototype,
+	}
+}
+
+func (node *StreamJoin) Get(ctx context.Context, variables octosql.Variables, streamID *StreamID) (RecordStream, *ExecutionOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+
+	// Get IDs of streams
+	leftSourceStreamID, err := GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakeString("left"))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get left source stream ID")
+	}
+
+	rightSourceStreamID, err := GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakeString("right"))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get right source stream ID")
+	}
+
+	leftStream, leftExec, err := node.leftSource.Get(ctx, variables, leftSourceStreamID)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get left source stream in stream join")
+	}
+
+	rightStream, rightExec, err := node.rightSource.Get(ctx, variables, rightSourceStreamID)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get right source stream in stream join")
+	}
+
+	// The source of watermarks for a stream join is the minimum of watermarks from its sources
+	watermarkSource := NewUnionWatermarkGenerator([]WatermarkSource{leftExec.WatermarkSource, rightExec.WatermarkSource})
+
+	trigger, err := node.triggerPrototype.Get(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get trigger from trigger prototype")
+	}
+
+	// Create the shuffles
+	mergedNextShuffles, err := mergeNextShuffles(leftExec.NextShuffles, rightExec.NextShuffles)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't merge next shuffles of sources")
+	}
+
+	stream := &JoinedStream{
+		streamID:       streamID,
+		eventTimeField: node.eventTimeField,
+		joinType:       node.joinType,
+	}
+
+	processFunc := &ProcessByKey{
+		eventTimeField:  node.eventTimeField,
+		trigger:         trigger,
+		keyExpressions:  [][]Expression{node.leftKey, node.rightKey},
+		processFunction: stream,
+		variables:       variables,
+	}
+
+	streamJoinPullEngine := NewPullEngine(processFunc, node.storage, []RecordStream{leftStream, rightStream}, streamID, watermarkSource, true)
+
+	return streamJoinPullEngine,
+		NewExecutionOutput(
+			streamJoinPullEngine,
+			mergedNextShuffles,
+			append(leftExec.TasksToRun, append(rightExec.TasksToRun, func() error { streamJoinPullEngine.Run(ctx); return nil })...),
+		), nil
+}
+
+type JoinedStream struct {
+	streamID       *StreamID
+	eventTimeField octosql.VariableName
+	joinType       JoinType
+}
+
+var leftStreamRecordsPrefix = []byte("$left_stream_records$")
+var rightStreamRecordsPrefix = []byte("$right_stream_records$")
+var matchesToTriggerPrefix = []byte("$matches_to_trigger$")
+var recordsToTriggerPrefix = []byte("$records_to_trigger$")
```

Comment:
I think that `singleRecordsToTriggerPrefix` might be a bit better name, since `records` can actually mean merged record so I got a bit confused (I know that in this scope joined records = matches but you know... :P )